engine/execution/ingestion/block_queue/queue.go

```diff
@@ -0,0 +1,382 @@
+}
+
+type MissingCollection struct {
+	BlockID   flow.Identifier
+	Height    uint64
+	Guarantee *flow.CollectionGuarantee
+}
+
+func (m *MissingCollection) ID() flow.Identifier {
+	return m.Guarantee.ID()
+}
+
+// collectionInfo is an internal struct used to keep track of the state of a collection,
+// and the blocks that include the collection
+type collectionInfo struct {
+	Collection *entity.CompleteCollection
+	IncludedIn map[flow.Identifier]*entity.ExecutableBlock
+}
+
+func NewBlockQueue(logger zerolog.Logger) *BlockQueue {
+	log := logger.With().Str("module", "block_queue").Logger()
+
+	return &BlockQueue{
+		log:              log,
+		blocks:           make(map[flow.Identifier]*entity.ExecutableBlock),
+		collections:      make(map[flow.Identifier]*collectionInfo),
+		blockIDsByHeight: make(map[uint64]map[flow.Identifier]*entity.ExecutableBlock),
+	}
+}
+
+// OnBlock is called when a new block is received, the parentFinalState indicates
+// whether its parent block has been executed.
+// Caller must ensure:
+// 1. blocks are passsed in order, i.e. parent block is passed in before its child block
+// 2. if a block's parent is not executed, then the parent block must be passed in first
+// 3. if a block's parent is executed, then the parent's finalState must be passed in
+// It returns (nil, nil, nil) if this block is a duplication
+func (q *BlockQueue) OnBlock(block *flow.Block, parentFinalState *flow.StateCommitment) (
+	[]*MissingCollection, // missing collections
+	[]*entity.ExecutableBlock, // blocks ready to execute
+	error, // exceptions
+) {
+	q.Lock()
+	defer q.Unlock()
+
+	// check if the block already exists
+	blockID := block.ID()
+	executable, ok := q.blocks[blockID]
+	// handle the case where the block has seen before
+	if ok {
+		// we have already received this block, and its parent still has not been executed yet
+		if executable.StartState == nil && parentFinalState == nil {
+			return nil, nil, nil
+		}
+
+		// this is an edge case where parentFinalState is provided, and its parent block exists
+		// in the queue but has not been marked as executed yet (OnBlockExecuted(parent) is not called),
+		// in this case, we will internally call OnBlockExecuted(parentBlockID, parentFinalState).
+		// there is no need to create the executable block again, since it's already created.
+		if executable.StartState == nil && parentFinalState != nil {
+			executables, err := q.onBlockExecuted(block.Header.ParentID, *parentFinalState)
+			if err != nil {
+				return nil, nil, fmt.Errorf("receiving block %v with parent commitment %v, but parent block %v already exists with no commitment, fail to call mark parent as executed: %w",
+					blockID, *parentFinalState, block.Header.ParentID, err)
+			}
+
+			// we already have this block, its collection must have been fetched, so we only return the
+			// executables from marking its parent as executed.
+			return nil, executables, nil
+		}
+
+		// this is an edge case could be ignored
```

Comment:
can you elaborate in this comment about how this case would happen and why it's safe to ignore